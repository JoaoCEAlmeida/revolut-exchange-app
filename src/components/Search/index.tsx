import React, { useCallback, useMemo, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { nanoid } from 'nanoid';

import { useAppDispatch } from '../../app/hooks';
import { setCurrencyFrom, setCurrencyTo } from '../../features/currencySlice';
import {
  SearchWrapper,
  StyledInput,
  InputWrapper,
  StyledButton,
  StyledButtonCurrency,
  NotFound
} from './styles';

const currencies = [
  { currency: 'EUR', id: 1 },
  { currency: 'USD', id: 2 },
  { currency: 'GBP', id: 3 }
];

const Search: React.FC = () => {
  const [searchTerm, setSearchTerm] = useState('');
  const queryString = useLocation().search;
  const navigate = useNavigate();

  const dispatch = useAppDispatch();

  const currencyType = new URLSearchParams(queryString).get('currencyType');

  // If API offered a search option we should use it
  // For now I'll use the 3 default currencies hard coded in the project.
  const filteredCurrency = useMemo(
    () =>
      currencies.filter((account) => {
        if (searchTerm == '') return account;
        else if (account.currency.toLocaleLowerCase().includes(searchTerm.toLocaleLowerCase()))
          return account;
      }),
    [searchTerm]
  );

  const goBack = useCallback(() => {
    navigate('/');
  }, [navigate]);

  const selectCurrency = useCallback(
    (currency: string) => () => {
      if (currencyType === 'currencyFrom') {
        dispatch(setCurrencyFrom(currency));
        navigate('/');
      } else if (currencyType === 'currencyTo') {
        dispatch(setCurrencyTo(currency));
        navigate('/');
      } else return;
    },
    [currencyType, dispatch, navigate]
  );
  const onChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    setSearchTerm(e.target.value);
  }, []);

  return (
    <SearchWrapper>
      <InputWrapper>
        <StyledInput type="text" placeholder="Search..." value={searchTerm} onChange={onChange} />
        <StyledButton text="Cancel" onClick={goBack} />
      </InputWrapper>
      {filteredCurrency.length ? (
        filteredCurrency.map((account) => {
          return (
            <StyledButtonCurrency
              key={nanoid()}
              text={account.currency}
              onClick={selectCurrency(account.currency)}
              data-testid="button"
            />
          );
        })
      ) : (
        <NotFound data-testid="not-found">The searched currency is not available.</NotFound>
      )}
    </SearchWrapper>
  );
};

export default Search;
