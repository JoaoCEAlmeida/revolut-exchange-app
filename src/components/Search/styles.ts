import styled from 'styled-components';
import { mobile } from '../../styles/mediaQueries';
import Input from '../../UI-kit/Input';
import Button from '../../UI-kit/Button';

export const SearchWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: 30px;
  ${mobile(`
    margin-left: 5px;
    margin-right: 5px;
  `)}
`;

export const InputWrapper = styled.div`
  display: flex;
  flex-direction: row;
  margin-bottom: 20px;
  justify-content: space-between;
`;

export const StyledInput = styled(Input)`
  display: flex;
  text-align: left;
  border-color: ${(props) => props.theme.colors.secondary.main};
  border-width: 0 0 2px;

  &:focus {
    outline: none;
    caret-color: ${(props) => props.theme.colors.primary.main};
    border-color: ${(props) => props.theme.colors.primary.main};
  }
`;

export const StyledButton = styled(Button)`
  && {
    width: 25%;
  }
`;

export const StyledButtonCurrency = styled(Button)`
  && {
    background-color: ${(props) => props.theme.colors.secondary.main};
    color: ${(props) => props.theme.colors.primary.main};
    width: 100px;
    margin-bottom: 10px;
  }
`;

export const NotFound = styled.span``;
