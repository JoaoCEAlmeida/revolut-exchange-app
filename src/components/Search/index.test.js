import renderWithProviders from '../../test/providers';
import { screen, fireEvent, act } from '@testing-library/react';
import Search from './index';

describe('SearchTest:', () => {
  test('should render correctly', () => {
    const { SearchTest } = renderWithProviders(() => <Search />);

    expect(SearchTest).toMatchSnapshot();
  });

  test('should show not found message', () => {
    renderWithProviders(() => <Search />);

    const searchInput = screen.getByPlaceholderText('Search...');

    act(() => {
      fireEvent.change(searchInput, { target: { value: 'asd' } });
    });

    expect(screen.getByTestId('not-found').textContent).toBe(
      'The searched currency is not available.'
    );
  });

  test('should filter accounts', () => {
    renderWithProviders(() => <Search />);

    const searchInput = screen.getByPlaceholderText('Search...');

    expect(screen.getAllByTestId('button')).toHaveLength(3);

    act(() => {
      fireEvent.change(searchInput, { target: { value: 'Gbp' } });
    });

    expect(screen.getAllByTestId('button')).toHaveLength(1);

    expect(screen.getByText('GBP')).toBeVisible();
  });
});
