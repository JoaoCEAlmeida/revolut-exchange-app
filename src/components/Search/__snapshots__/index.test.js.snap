// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeFormTest: should render correctly 1`] = `undefined`;

exports[`SearchTest: should render correctly 1`] = `undefined`;
